signer/src/bitcoin/validation.rs

```diff
@@ -0,0 +1,434 @@
+    /// The deposit UTXO has already been spent.
+    #[error("deposit transaction used as input in confirmed sweep transaction; {0}")]
+    DepositUtxoSpent(OutPoint),
+    /// Given the current time and block height, it would be imprudent to
+    /// attempt to sweep in a deposit request with the given lock-time.
+    #[error("lock-time expiration is too soon; {0}")]
+    LockTimeExpiry(OutPoint),
+    /// The signer does not have a record of their vote on the deposit
+    /// request in their database.
+    #[error("the signer does not have a record of their vote on the deposit request; {0}")]
+    NoVote(OutPoint),
+    /// The signer has rejected the deposit request.
+    #[error("the signer has not accepted the deposit request; {0}")]
+    RejectedRequest(OutPoint),
+    /// The signer does not have a record of the deposit request in their
+    /// database.
+    #[error("the signer does not have a record of the deposit request; {0}")]
+    Unknown(OutPoint),
+    /// The locktime in the reclaim script is in time units and that is not
+    /// supported. This shouldn't happen, since we will not put it in our
+    /// database is this is the case.
+    #[error("the deposit locktime is denoted in time and that is not supported; {0}")]
+    UnsupportedLockTime(OutPoint),
+}
+
+/// The responses for validation of a sweep transaction on bitcoin.
+#[derive(Debug, thiserror::Error, PartialEq, Eq, Copy, Clone)]
+pub enum BitcoinSweepErrorMsg {
+    /// The error has something to do with the inputs.
+    #[error("the assessed fee for a deposit would exceed their max-fee")]
+    Deposit(#[from] BitcoinDepositInputError),
+}
+
+/// A struct for a bitcoin validation error containing all the necessary
+/// context.
+#[derive(Debug)]
+pub struct BitcoinValidationError {
+    /// The specific error that happened during validation.
+    pub error: BitcoinSweepErrorMsg,
+    /// The additional information that was used when trying to
+    /// validate the complete-deposit contract call. This includes the
+    /// public key of the signer that was attempting to generate the
+    /// `complete-deposit` transaction.
+    pub context: BitcoinTxContext,
+}
+
+impl std::fmt::Display for BitcoinValidationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // TODO(191): Add the other variables to the error message.
+        self.error.fmt(f)
+    }
+}
+
+impl std::error::Error for BitcoinValidationError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.error)
+    }
+}
+
+/// An enum for the confirmation status of a deposit request.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DepositRequestStatus {
+    /// We have a record of the deposit request transaction, and it has
+    /// been confirmed on the canonical bitcoin blockchain. We have not
+    /// spent these funds. The integer is the height of the block
+    /// confirming the deposit request.
+    Confirmed(u64),
```

Comment:
Wouldn't it be better to have the combination of hash + height since the report doesn't have the hash and we can have multiple block hashes for the same height? Not sure if it'd necessarily be used, but can be useful for logging.

Reply:
Hmmm, yeah I think you're right. Will do.

Reply:
Updated in https://github.com/stacks-network/sbtc/pull/744/commits/1da998d1d1db6b789b67ee90d1ff08d57a170a0e